pkg/volume/aws_ebs/aws_util.go

```diff
@@ -86,68 +87,91 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
+		glog.Errorf("error building labels for new EBS volume %q: %v", name, err)
+	}
+
+	fstype := ""
+	if v, ok := c.options.Parameters[volume.VolumeParameterFSType]; ok {
+		fstype = v
+	}
+
+	return name, int(requestGB), labels, fstype, nil
+}
+
+// returns volumeOptions for EBS based on storageclass parameters and node configuration
+func populateVolumeOptions(pluginName, pvcName string, capacityGB int, tags map[string]string, storageParams map[string]string, node *v1.Node, allowedTopologies []v1.TopologySelectorTerm) (*aws.VolumeOptions, error) {
 	volumeOptions := &aws.VolumeOptions{
-		CapacityGB: requestGB,
+		CapacityGB: capacityGB,
 		Tags:       tags,
-		PVCName:    c.options.PVC.Name,
+		PVCName:    pvcName,
 	}
-	fstype := ""
+
 	// Apply Parameters (case-insensitive). We leave validation of
 	// the values to the cloud provider.
-	volumeOptions.ZonePresent = false
-	volumeOptions.ZonesPresent = false
-	for k, v := range c.options.Parameters {
+	zonePresent := false
+	zonesPresent := false
+	var zone, zones string
+	var err error
+	for k, v := range storageParams {
 		switch strings.ToLower(k) {
 		case "type":
 			volumeOptions.VolumeType = v
 		case "zone":
-			volumeOptions.ZonePresent = true
-			volumeOptions.AvailabilityZone = v
+			zonePresent = true
+			zone = v
 		case "zones":
-			volumeOptions.ZonesPresent = true
-			volumeOptions.AvailabilityZones = v
+			zonesPresent = true
+			zones = v
 		case "iopspergb":
 			volumeOptions.IOPSPerGB, err = strconv.Atoi(v)
 			if err != nil {
-				return "", 0, nil, "", fmt.Errorf("invalid iopsPerGB value %q, must be integer between 1 and 30: %v", v, err)
+				return nil, fmt.Errorf("invalid iopsPerGB value %q, must be integer between 1 and 30: %v", v, err)
 			}
 		case "encrypted":
 			volumeOptions.Encrypted, err = strconv.ParseBool(v)
 			if err != nil {
-				return "", 0, nil, "", fmt.Errorf("invalid encrypted boolean value %q, must be true or false: %v", v, err)
+				return nil, fmt.Errorf("invalid encrypted boolean value %q, must be true or false: %v", v, err)
 			}
 		case "kmskeyid":
 			volumeOptions.KmsKeyId = v
-		case volume.VolumeParameterFSType:
-			fstype = v
 		default:
-			return "", 0, nil, "", fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, pluginName)
 		}
 	}
 
-	if volumeOptions.ZonePresent && volumeOptions.ZonesPresent {
-		return "", 0, nil, "", fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
-	}
-
-	// TODO: implement PVC.Selector parsing
-	if c.options.PVC.Spec.Selector != nil {
-		return "", 0, nil, "", fmt.Errorf("claim.Spec.Selector is not supported for dynamic provisioning on AWS")
-	}
-
-	name, err := cloud.CreateDisk(volumeOptions)
+	zoneCfg, err := volumeutil.PopulateZoneConfig(zonePresent, zonesPresent, zone, zones, node, ebsMaxReplicasInAZ, allowedTopologies)
```

Comment:
The aws cloud provider interface abstraction seems a bit off to me.  I feel like the aws cloud provider layer should not be processing the `zone` and `zones` arguments, which are StorageClass parameters.  It seems better suited for the Kubernetes plugin layer to process, and pass in the selected zone to the aws cloud provider layer.  Right now, both layers have to handle all the permutations of `zone` and `zones`.

If the aws cloud layer is refactored to push the zone argument processing up to the aws plugin, then I think this `PopulateZoneConfig` could be greatly simplified, and only needs to return a list of zones, either from allowedTopologies or from old `zone/zones`. @jsafrane wdyt?